Makefile

```diff
@@ -90,6 +110,80 @@ test-silent:
 benchmark-test:
 	. ./scripts/shared_env && go test -run=XX -bench=. $(shell go list ./agent/... | grep -v /vendor/)
 
+define dockerbuild
+	docker run \
+		--net none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+define win-cgo-dockerbuild
+	docker run --net=none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--env "GOOS=windows" \
+		--env "CGO_ENABLED=1" \
+		--env "CC=x86_64-w64-mingw32-gcc" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+# TODO: use `go list -f` to target the test files more directly
+ALL_GO_FILES = $(shell find . -name "*.go" -print | tr "\n" " ")
+GO_INTEG_TEST = go test -race -tags integration -c -o
+out/test-artifacts/linux-engine-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/linux-app-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/linux-stats-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+out/test-artifacts/windows-engine-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/windows-app-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/windows-stats-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+GO_FUNCTIONAL_TEST = go test -tags functional -c -o
+out/test-artifacts/linux-simple-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests/generated/simpletests_unix)
+
+out/test-artifacts/linux-handwritten-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests)
```

Comment:
this is not `./agent/functional_tests/tests/...` because ?

Reply:
The `-c` test flag only works on a single package. Also, the only test files it needs to be aware of are in this folder.